runtime/libc/wasi/wasi_impl_minimal.c

```diff
@@ -0,0 +1,1129 @@
+    /* similar to `futimens` in POSIX. */
+    return wasi_unsupported_syscall(__func__);
+}
+
+/**
+ * Read from a file descriptor without updating the descriptor's offset
+ *
+ * @param fd
+ * @param iovs_baseptr List of scatter/gather vectors in which to store data.
+ * @param iovs_len The length of the array pointed to by `iovs`.
+ * @param offset The offset within the file at which to read.
+ * @param nread_retptr The number of bytes read.
+ * @return status code
+ */
+__wasi_errno_t __wasi_fd_pread(void* wasi_context, __wasi_fd_t fd, const __wasi_iovec_t* iovs, size_t iovs_len,
+                               __wasi_filesize_t offset, __wasi_size_t* nread_retptr) {
+    /* similar to `preadv` in POSIX. */
+    return wasi_unsupported_syscall(__func__);
+}
+
+/**
+ * Return a description of the given preopened file descriptor.
+ *
+ * @param fd
+ * @param prestat_retptr The buffer where the description is stored.
+ * @return status code
+ */
+__wasi_errno_t __wasi_fd_prestat_get(void* wasi_context, __wasi_fd_t fd, __wasi_prestat_t* prestat_retptr) {
+    return wasi_unsupported_syscall(__func__);
+}
+
+/**
+ * Return a description of the given preopened file descriptor.
+ *
+ * @param fd
+ * @param path_retptr A buffer into which to write the preopened directory name.
+ * @param path_len The length of the buffer at path_retptr
+ * @return status code
+ */
+__wasi_errno_t __wasi_fd_prestat_dir_name(void* wasi_context, __wasi_fd_t fd, char* path, __wasi_size_t path_len) {
+    return wasi_unsupported_syscall(__func__);
+}
+
+/**
+ * Write to a file descriptor without updating the descriptor's offset
+ *
+ * @param fd
+ * @param iovs_baseptr List of scatter/gather vectors from which to retrieve data.
+ * @param iovs_len The length of the array pointed to by `iovs`.
+ * @param offset The offset within the file at which to write.
+ * @param nwritten_retptr The number of bytes written.
+ * @return status code
+ *
+ */
+__wasi_errno_t __wasi_fd_pwrite(void* wasi_context, __wasi_fd_t fd, const __wasi_ciovec_t* iovs, size_t iovs_len,
+                                __wasi_filesize_t offset, __wasi_size_t* nwritten_retptr) {
+    /* similar to `pwritev` in POSIX. */
+    return wasi_unsupported_syscall(__func__);
+}
+
+/**
+ * Read from a file descriptor
+ *
+ * @param fd
+ * @param iovs_baseptr
+ * @param iovs_len
+ * @param nwritten_retptr The number of bytes read.
+ * @return __WASI_ERRNO_SUCCESS, WASI_EAGAIN, WASI_EWOULDBLOCK, WASI_EBADF, WASI_EFAULT, WASI_EINTR, WASI_EIO,
+ * WASI_EISDIR, or others
+ */
+__wasi_errno_t __wasi_fd_read(void* wasi_context, __wasi_fd_t fd, const __wasi_iovec_t* iovs, size_t iovs_len,
+                              __wasi_size_t* nwritten_retptr) {
+    /* TODO: Should this use readv instead of read? */
+    __wasi_size_t sum = 0;
+
+    for (int i = 0; i < iovs_len; i++) {
+        ssize_t res = read(fd, iovs[i].buf, iovs[i].buf_len);
```

Comment:
This file isn't a minimal example if it relies on POSIX. Implies that we'll have to implement a separate microcontroller version, which is OK.

Reply:
Make sure that Maorui knows this.

Reply:
Added README documentation and started a discussion with him around this on Slack. 🙌